src/wolfkrow/core/engine/resolver.py

```diff
@@ -1,40 +1,163 @@
 import os
 import re
 import string
 
-class ReplacementsDict(dict):
-    """ Implement the __missing__ method for dictionary so that it returns '{<key>}' 
-        when the value is not present so that when it is passed to the str.format 
-        function, missing keys do not raise an exception.
-    """
+from string import Formatter
 
-    def __missing__(self, key):
-        return "{" + key + "}"
+class WolfkrowFormatter(Formatter):
+    def _vformat(self, format_string, args, kwargs, used_args, recursion_depth, auto_arg_index=0):
+        """
+            Overwrite the _vformat method of the Formatter class to allow for optional
+            {} style replacements. If the replacement is not found in the replacements
+            then it will simply be kept as is.
+        """
+        if recursion_depth < 0:
+            raise ValueError('Max string recursion exceeded')
+        result = []
+        for literal_text, field_name, format_spec, conversion in \
+                self.parse(format_string):
+
+            # output the literal text
+            if literal_text:
+                result.append(literal_text)
+
+            # if there's a field, output it
+            if field_name is not None:
+                # this is some markup, find the object and do
+                #  the formatting
+
+                # handle arg indexing when empty field_names are given.
+                if field_name == '':
+                    if auto_arg_index is False:
+                        raise ValueError('cannot switch from manual field '
+                                         'specification to automatic field '
+                                         'numbering')
+                    field_name = str(auto_arg_index)
+                    auto_arg_index += 1
+                elif field_name.isdigit():
+                    if auto_arg_index:
+                        raise ValueError('cannot switch from manual field '
+                                         'specification to automatic field '
+                                         'numbering')
+                    # disable auto arg incrementing, if it gets
+                    # used later on, then an exception will be raised
+                    auto_arg_index = False
+
+                # given the field_name, find the object it references
+                #  and the argument it came from
+
+                # WOLFKROW CUSTOMIZATION: If the field_name is not found in the 
+                # kwargs or args, then just return the original string
+                try:
+                    obj, arg_used = self.get_field(field_name, args, kwargs)
+                except (KeyError, IndexError):
+                    obj = "{" + field_name
+                    if format_spec:
+                        obj += ":" + format_spec
+                    
+                    if conversion:
+                        obj += conversion
+
+                    obj += "}"
+                    result.append(obj)
```

Comment:
Lets re-raise a custom Wolfkrow Replacement Exception here with a meaningful error message. Then we can make the resolver catch it and print a warning.

Reply:
(Now that the resolver only tries to resolve things it actually needs)